File: Programmers/Lv3/섬 연결하기.py
def solution(n, costs):
    answer = 0
    costs.sort(key=lambda x: x[2])
    # weight 기준 오름차순 정렬
    # 알아서 낮은거 부터 들어감

    connect = set([costs[0][0]])
    # 간선 연결 정보

    while len(connect) != n:
        for cost in costs:
            v, w = cost[0], cost[1]
            weigth = cost[2]

            if v in connect and w in connect:
                continue
            if v in connect or w in connect:
                connect.update([cost[0], cost[1]])
                answer += weigth
                break
    # 딱 푸는게 그리디인데 그래프다?
    # Kruskal MST
    return answer

File: Programmers/Lv3/섬 연결하기.js
function solution(n, costs) {
  var graph = Array.from(Array(n), () => []);
  for (let [a, b, w] of costs) {
    graph[a].push([b, w]);
    graph[b].push([a, w]);
  }

  var visited = new Array(n).fill(false);
  const pq = [[0, 0]];
  // cost, nodeNum

  let total = 0;
  let edgeCnt = 0;

  while (pq.length > 0 && edgeCnt < n) {
    pq.sort((a, b) => a[0] - b[0]);
    const [cost, node] = pq.shift();

    if (visited[node]) continue;

    visited[node] = true;
    total += cost;
    edgeCnt++;

    for (let [nxtNode, nxtCost] of graph[node]) {
      if (!visited[nxtNode]) {
        pq.push([nxtCost, nxtNode]);
      }
    }
  }

  return total;
}
